// darwin-developer/lib/security/listKeychains.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _teen_process = require('teen_process');

var DOMAIN = {
    USER: 'user',
    SYSTEM: 'system',
    COMMON: 'common',
    DYNAMIC: 'dynamic'
};

exports.DOMAIN = DOMAIN;

exports['default'] = function callee$0$0() {
    var domain = arguments.length <= 0 || arguments[0] === undefined ? null : arguments[0];

    var domainArgs, _ref, stdout;

    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                domainArgs = domain ? ['-d', domain] : [];
                context$1$0.next = 3;
                return _regeneratorRuntime.awrap((0, _teen_process.exec)('security', ['list-keychains'].concat(domainArgs)));

            case 3:
                _ref = context$1$0.sent;
                stdout = _ref.stdout;
                return context$1$0.abrupt('return', stdout
                // Split into single lines
                .split('\n')
                // Extract keychains
                .map(function (line) {
                    return line.replace(/\s+"(.+)"\s*$/, '$1');
                }));

            case 6:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
};
